### PR TITLE
[WIP] ANI pre-computation

### PR DESCRIPTION
**Pull request by maxentile**
## Description

A performance bottleneck in our experiments with re-fitting ANI to free energy data has been to repeatedly compute dU/dtheta on a large collection of static snapshots x that doesn't change from iteration to iteration, where theta is a small subset of model parameters that governs only the final layer in a neural network inside the definition of U. A lot of expensive parts of that computation will be identical across all iterations. Can we split those out?

In this PR, we re-write U(x, theta) = g(f(x), theta), where f doesn't depend on the parameters theta we are varying, and where g is much cheaper than evaluating U from scratch.

Specifically, we define f to be all of the steps up until the last neural network layer (computing symmetry functions, then computing all neural network layers up until the last one), and we define g to be the very last layer.

Precomputing y = f(x) and only re-computing g(y) could potentially speed up our free-energy-fitting parameter optimizations quite a bit. The cost of recomputing U from scratch on a batch of 1000 methane configurations is over 1 second, but the cost of computing just the last layer g is ~40 milliseconds.

```
coordinates.shape torch.Size([1000, 5, 3])
time to compute energies in batch: 1.227 s
time to compute (dU / d all_params) in batch: 1.353 s
energy error: |g(f(x)) - model(x)|:  0.0
time to precompute up until last layer (f): 1.271 s
time to compute last layer (g): 0.038 s
time to compute derivatives of E w.r.t. last layer: 0.045 s
```

(Those timings were measured on a CPU instance in this Colab notebook: [Open In Colab](https://colab.research.google.com/drive/18N7quYsraPNpCS-JYUZxRwyil8NFbv7_?usp=sharing) )

(Thanks to @jaimergp and @wiederm for a helpful discussion this week that led me to try this.)

## Questions
- [ ] Should we generalize this to "checkpoint" somewhere else in the computation? (For our purposes, is it safe to assume we only ever want to touch the last layer?)
- [ ] Here I hardcoded the assumption that the last layer dimension is 160 -- this is true for the ANI-2x networks for carbon and hydrogen, but invalid for the networks for the other elements (for nitrogen and oxygen networks the last layer dimension is 128, and for sulfur, fluorine, and chlorine the last layer dimension is 96). How best to handle this "jagged" computation? (Two options come to mind: (1) emulate jagged array using sufficiently large padded array, (2) instead of returning/accepting a single array in intermediate computation, return a dictionary mapping element symbols to possibly different-sized arrays. I'm leaning towards option (2).)
- [ ] If further speed-up is needed, does g play nice with torchscript?
- [ ] This seems like something other PyTorchers must have wanted to do in other transfer-learning-y settings -- is there a more concise and maintainable way to do this checkpointing, that doesn't require reaching so far into the guts of the model?

## Status
- [x] Working for ANI-2x element-specific neural networks with last-layer dimension=160 (C, H)
- [ ] Working for all the other elements (N, O, S, F, Cl)
- [ ] Performance measured for droplet-sized systems
- [ ] Incorporated into AlchemicalANI and friends
- [ ] Incorporated into overall free-energy-fitting workflow
- [ ] Louder warning labels affixed to brittle dimension assumptions

### CODE REVIEW

**Comment by maxentile** [comment]
I had mistakenly branched this from `dev-mw` rather than from `master`, making this PR look much more daunting than it should! Closing this and opening a new PR with just the pre-computation script.
